### PR TITLE
Stream Multi Receive implementation

### PR DESCRIPTION
**Pull request by ami-GS**
## Description

Stream Multi Receive internal feature.
Dispatched from https://github.com/microsoft/msquic/pull/4182

## Testing

RecvBufferTest.cpp

## Documentation

N/A

### CODE REVIEW

**Comment by nibanks** [comment]
Just include the recv buffer and test changes? Not everything else

**Comment by ami-GS** [comment]
settings.c and stream.c etc. are for public api?

File: src/core/recv_buffer.h [hunk 1]
```diff
@@ -20,7 +20,7 @@ typedef enum QUIC_RECV_BUF_MODE {
 //
 typedef struct QUIC_RECV_CHUNK {
     CXPLAT_LIST_ENTRY Link;         // Link in the list of chunks.
-    uint32_t AllocLength : 31;      // Allocation size of Buffer
+    uint32_t AllocLength : 31;      // Allocation size of Buffer,
```
Comment (nibanks) [inline src/core/recv_buffer.h:23]: nit: revert

File: src/core/recv_buffer.h [hunk 2]
```diff
@@ -68,6 +68,12 @@ typedef struct QUIC_RECV_BUFFER {
     //
     uint32_t VirtualBufferLength;
 
+    //
+    // Basically same as Chunk->AllocLength of first chunk,
+    // but start shrinking by drain operation after next chunk is allocated.
```
Comment (nibanks) [inline src/core/recv_buffer.h:73]: nit: please fill up first line (to 80 column) before wrap

File: src/core/unittest/RecvBufferTest.cpp [hunk 3]
```diff
@@ -59,6 +59,7 @@ struct RecvBuffer {
         _Out_ BOOLEAN* NewDataReady
         ) {
         auto BufferToWrite = new (std::nothrow) uint8_t[WriteLength];
+        memset(BufferToWrite, 0, WriteLength); // Zero out the buffer (for debugging purposes
```
Comment (nibanks) [inline src/core/unittest/RecvBufferTest.cpp:62]:
```suggestion
        memset(BufferToWrite, 0, WriteLength); // Zero out the buffer (for debugging purposes)
```

File: src/core/unittest/RecvBufferTest.cpp [hunk 4]
```diff
@@ -560,8 +563,14 @@ TEST_P(WithMode, ReadPendingMultiWrite)
     RecvBuf.Read(&ReadOffset, &BufferCount, ReadBuffers);
     ASSERT_FALSE(RecvBuf.HasUnreadData());
     ASSERT_EQ(32ull, ReadOffset);
-    ASSERT_EQ(1u, BufferCount);
-    ASSERT_EQ(48u, ReadBuffers[0].Length);
+    if (Mode == QUIC_RECV_BUF_MODE_MULTIPLE) {
```
Comment (nibanks) [inline src/core/unittest/RecvBufferTest.cpp:566]: Why is this needed?

File: src/core/unittest/RecvBufferTest.cpp [hunk 5]
```diff
@@ -693,6 +702,767 @@ TEST_P(WithMode, DrainFrontChunkWithPendingGap)
     ASSERT_TRUE(RecvBuf.Drain(1));
 }
 
+void Check(
```
Comment (nibanks) [inline src/core/unittest/RecvBufferTest.cpp:705]: Move these functions into RecvBuffer

File: src/core/unittest/RecvBufferTest.cpp (hunk 4 above)
Comment (ami-GS) [reply src/core/unittest/RecvBufferTest.cpp:566]: Default AllocLength is 64. Then

Write 0,32
Read
Write 32,48 (append next chunk)
Drain 32
Read -> 32 from first chunk, 16 from second chunk.

File: src/core/recv_buffer.c [hunk 6]
```diff
@@ -419,58 +421,100 @@ QuicRecvBufferCopyIntoChunks(
                 RecvBuffer->Chunks.Flink, // First chunk
                 QUIC_RECV_CHUNK,
                 Link);
-        uint32_t ChunkLength;
+
+        uint64_t ChunkLength;
```
Comment (nibanks) [inline src/core/recv_buffer.c:425]: Does this really need to be 64 bits? We only use 31 bits for the allocation length of a chunk. So it should be impossible for ChunkLength to be bigger, right?

**Comment by ami-GS** [comment]
@nibanks This PR is ready to go. there is stall issue, but it is with API implementation

**Review by nibanks — APPROVED** [review]
